### PR TITLE
tcbinfo:remove total_num form tcbinfo.

### PR DESCRIPTION
**Pull request by anjiahao1**
## Summary
tcbinfo:remove total_num form tcbinfo. Can solve Mac os alignment problem
## Impact
debug with jlink
## Testing
`make -f tools/Makefile.host`
`cp tools/jlink-nuttx /opt/SEGGER/JLink_V786a/libnuttxplugin.so `
`JLinkGDBServer -if SWD -speed 5000 -device STM32F429ZI -NoGui 1 -rtos libnuttxplugin`

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@duduita please try this patch.

File: include/nuttx/sched.h [hunk 1]
```diff
@@ -715,8 +715,7 @@ begin_packed_struct struct tcbinfo_s
   uint16_t stack_off;                    /* Offset of tcb.stack_alloc_ptr   */
   uint16_t stack_size_off;               /* Offset of tcb.adj_stack_size    */
   uint16_t regs_off;                     /* Offset of tcb.regs              */
-  uint16_t basic_num;                    /* Num of genernal regs            */
-  uint16_t total_num;                    /* Num of regs in tcbinfo.reg_offs */
+  uint16_t regs_num;                     /* Num of genernal regs            */
```
Comment (acassis) [inline include/nuttx/sched.h:718]:
```suggestion
  uint16_t regs_num;                     /* Num of general regs            */

**Comment by acassis** [comment]
@anjiahao1 I noticed that you also removed FPU regs and some other regs from TCB, could you please inclusive more info about it in the commit message?

**Comment by xiaoxiang781216** [comment]
> @anjiahao1 I noticed that you also removed FPU regs and some other regs from TCB, could you please inclusive more info about it in the commit message?

since gdb doesn't display FPU register, the information isn't used at all as @anjiahao1 mention in the commit message and pr description.

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@anjiahao1 

>tcbinfo:remove total_num form tcbinfo. Can solve Mac os alignment problem

Did you test this PR on the M1/M2 Mac, as reported in https://github.com/apache/nuttx/issues/11087

**Comment by xiaoxiang781216** [comment]
> @anjiahao1
> 
> > tcbinfo:remove total_num form tcbinfo. Can solve Mac os alignment problem
> 
> Did you test this PR on the M1/M2 Mac, as reported in #11087

We don't have M1/M2 Mac to try this patch, the fix base on the link error:
```
ld: warning: pointer not aligned at address 0x10006C5D2 ('_g_tcbinfo' + 18 from nuttx.rel)
ld: unaligned pointer(s) for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
18 is the offset of reg_off field in old tcbinfo_s. With this patch, the offset change to 16 which should satisfy the alignment requirement.

**Comment by masayuki2009** [comment]
@xiaoxiang781216 @anjiahao1 

>We don't have M1/M2 Mac to try this patch, the fix base on the link error:

Hmm, that should be described in the summary, because

>tcbinfo:remove total_num form tcbinfo. Can solve Mac os alignment problem

File: arch/arm/src/armv6-m/arm_tcbinfo.c [hunk 2]
```diff
@@ -51,13 +51,14 @@ static const uint16_t g_reg_offs[] =
   TCB_REG_OFF(REG_R14),
   TCB_REG_OFF(REG_R15),
   TCB_REG_OFF(REG_XPSR),
-
+#if 0
```
Comment (pkarashchenko) [inline arch/arm/src/armv6-m/arm_tcbinfo.c:54]: Why do we need to keep all the `#if 0` stuff?

**Comment by anjiahao1** [comment]
> I noticed that you also removed FPU regs and some other regs from TCB, could you please inclusive more info about it in the commit message?

In most cases, gdb does not need fpu registers

**Comment by masayuki2009** [comment]
@anjiahao1 

As long as I built this PR on  M1 MBP, 

```
CC:  sim/sim_tcbinfo.c
cc -c -g -fomit-frame-pointer -fno-common -fvisibility=hidden -ffunction-sections -fdata-sections -Wall -Wstrict-prototypes -Wshadow -Wundef -Wno-attributes -Wno-uno
sim/sim_tcbinfo.c:96:28: error: use of undeclared identifier 'g_reg_offs'
  .regs_num       = nitems(g_reg_offs),
                           ^
sim/sim_tcbinfo.c:96:28: error: use of undeclared identifier 'g_reg_offs'
sim/sim_tcbinfo.c:98:10: error: use of undeclared identifier 'g_reg_offs'
    .p = g_reg_offs,
```

My environment is

```
$ sw_vers
ProductName:		macOS
ProductVersion:		13.6
BuildVersion:		22G120

$ cc --version
Apple clang version 15.0.0 (clang-1500.0.40.1)
Target: arm64-apple-darwin22.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by anjiahao1** [comment]
> g_reg_offs

This should be because mac arm does not implement g_reg_offs. I will fix it later.

**Comment by anjiahao1** [comment]
> @anjiahao1
> 
> As long as I built this PR on M1 MBP,
> 
> ```
> CC:  sim/sim_tcbinfo.c
> cc -c -g -fomit-frame-pointer -fno-common -fvisibility=hidden -ffunction-sections -fdata-sections -Wall -Wstrict-prototypes -Wshadow -Wundef -Wno-attributes -Wno-uno
> sim/sim_tcbinfo.c:96:28: error: use of undeclared identifier 'g_reg_offs'
>   .regs_num       = nitems(g_reg_offs),
>                            ^
> sim/sim_tcbinfo.c:96:28: error: use of undeclared identifier 'g_reg_offs'
> sim/sim_tcbinfo.c:98:10: error: use of undeclared identifier 'g_reg_offs'
>     .p = g_reg_offs,
> ```
> 
> My environment is
> 
> ```
> $ sw_vers
> ProductName:		macOS
> ProductVersion:		13.6
> BuildVersion:		22G120
> 
> $ cc --version
> Apple clang version 15.0.0 (clang-1500.0.40.1)
> Target: arm64-apple-darwin22.6.0
> Thread model: posix
> InstalledDir: /Library/Developer/CommandLineTools/usr/bin
> ```

This should be because mac arm does not implement g_reg_offs. I will fix it later.

**Review by jerpelea — APPROVED** [review]

**Comment by anjiahao1** [comment]
@masayuki2009  i push newset patch，and test on qemu arm64 run ubuntu with arm64 sim.
```
ajh@arm64-mikasa:~/work/nuttx$ gdb-multiarch nuttx -ex "source tools/gdb/__init__.py"
GNU gdb (Ubuntu 13.1-2ubuntu2) 13.1
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "aarch64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from nuttx...
set pagination off
source /home/ajh/work/nuttx/tools/gdb/memdump.py
source /home/ajh/work/nuttx/tools/gdb/lists.py

if use thread command, please don't use 'continue', use 'c' instead !!!
source /home/ajh/work/nuttx/tools/gdb/thread.py
source /home/ajh/work/nuttx/tools/gdb/utils.py
"handle SIGUSR1 "nostop" "pass" "noprint"
(gdb) r
Starting program: /home/ajh/work/nuttx/nuttx
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".

NuttShell (NSH) NuttX-10.1.0
MOTD: username=admin password=Administrator
nsh>
Program received signal SIGSTOP, Stopped (signal).
0x0000fffff7e831e8 in __GI___clock_nanosleep (clock_id=<optimized out>, clock_id@entry=0, flags=flags@entry=0, req=req@entry=0xfffffffff048, rem=rem@entry=0x0) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:78
78      ../sysdeps/unix/sysv/linux/clock_nanosleep.c: No such file or directory.
(gdb) info threads
Id   Thread                Info                                                                             Frame
*0   Thread 0x4703d0       (Name: Idle_Task, State: Running, Priority: 0, Stack: 69616)                     0xfffff7e831e8 clock_nanosleep() at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:78
1    Thread 0xfffff3ed14f0 (Name: loop_task, State: Waiting,Signal, Priority: 224, Stack: 67536)            0x413530 up_switch_context() at sim/sim_switchcontext.c:94
2    Thread 0xfffff3ee2630 (Name: hpwork, State: Waiting,Semaphore, Priority: 224, Stack: 67504)            0x413530 up_switch_context() at sim/sim_switchcontext.c:94
3    Thread 0xfffff3ef3560 (Name: nsh_main, State: Waiting,Semaphore, Priority: 100, Stack: 67536)          0x413530 up_switch_context() at sim/sim_switchcontext.c:94
(gdb) thread 2
(gdb) bt
#0  0x0000000000413530 in up_switch_context (tcb=0xfffff3ed14f0, rtcb=0xfffff3ee2630) at sim/sim_switchcontext.c:94
#1  0x0000000000406804 in nxsem_wait (sem=0x470ec0 <g_hpwork+16>) at semaphore/sem_wait.c:176
#2  0x0000000000406850 in nxsem_wait_uninterruptible (sem=0x470ec0 <g_hpwork+16>) at semaphore/sem_wait.c:248
#3  0x0000000000405cf8 in work_thread (argc=3, argv=0xfffff3ee2d60) at wqueue/kwork_thread.c:207
#4  0x00000000004076f0 in nxtask_start () at task/task_start.c:129
#5  0x0000000000413098 in pre_start () at sim/sim_initialstate.c:52
#6  0x0000000000413070 in up_getsp () at /home/ajh/work/nuttx/include/arch/irq.h:138
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

look like run well.

**Comment by masayuki2009** [comment]
@anjiahao1 

>@masayuki2009 i push newset patch，and test on qemu arm64 run ubuntu with arm64 sim.

It seems that the build error has been resolved.
However, it crashed.

```
$ ./nuttx
zsh: segmentation fault  ./nuttx
```

**Comment by anjiahao1** [comment]
> @anjiahao1
> 
> > @masayuki2009 i push newset patch，and test on qemu arm64 run ubuntu with arm64 sim.
> 
> It seems that the build error has been resolved. However, it crashed.
> 
> ```
> $ ./nuttx
> zsh: segmentation fault  ./nuttx
> ```

What configuration is running? I am using sim:nsh.

**Comment by masayuki2009** [comment]
@anjiahao1 

>What configuration is running? I am using sim:nsh.

I tried `sim:nsh` but the same crash also happened with `sim:smp`.

**Comment by anjiahao1** [comment]
@masayuki2009 
i used mac os get this
image
This does not seem to be caused by this patch. I am not very familiar with how to debug on macos. The backtace function of lldb does not seem to work.

**Comment by anjiahao1** [comment]
@masayuki2009 
```
(base) ➜  nuttx git:(master) ✗ sudo lldb nuttx
(lldb) target create "nuttx"
Current executable set to '/Users/pengxianghao/Documents/code/nuttx_community/nuttx/nuttx' (arm64).
(lldb) b nx_start
Breakpoint 1: where = nuttx`nx_start, address = 0x0000000100006434
(lldb) c
error: Command requires a current process.
(lldb) r
Process 24321 launched: '/Users/pengxianghao/Documents/code/nuttx_community/nuttx/nuttx' (arm64)
Process 24321 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x0000000100006434 nuttx`nx_start
nuttx`nx_start:
->  0x100006434 <+0>:  sub    sp, sp, #0x70
    0x100006438 <+4>:  stp    x29, x30, [sp, #0x60]
    0x10000643c <+8>:  b      0x100006440               ; <+12>
    0x100006440 <+12>: b      0x100006444               ; <+16>
Target 0: (nuttx) stopped.
(lldb) b bzero
Breakpoint 2: 4 locations.
(lldb) c
Process 24321 resuming
Process 24321 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 2.1
    frame #0: 0x0000000100024584 nuttx`bzero
nuttx`bzero:
->  0x100024584 <+0>:  sub    sp, sp, #0x20
    0x100024588 <+4>:  stp    x29, x30, [sp, #0x10]
    0x10002458c <+8>:  str    x0, [sp, #0x8]
    0x100024590 <+12>: str    x1, [sp]
Target 0: (nuttx) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 2.1
  * frame #0: 0x0000000100024584 nuttx`bzero
    frame #1: 0x0000000100006488 nuttx`nx_start + 84
    frame #2: 0x00000001000042f4 nuttx`sim_gpio_initialize - 18446744069414567179
(lldb) c
Process 24321 resuming
Process 24321 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 2.1
    frame #0: 0x0000000100024584 nuttx`bzero
nuttx`bzero:
->  0x100024584 <+0>:  sub    sp, sp, #0x20
    0x100024588 <+4>:  stp    x29, x30, [sp, #0x10]
    0x10002458c <+8>:  str    x0, [sp, #0x8]
    0x100024590 <+12>: str    x1, [sp]
Target 0: (nuttx) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 2.1
  * frame #0: 0x0000000100024584 nuttx`bzero
    frame #1: 0x00000001000245a0 nuttx`bzero + 28
    frame #2: 0x0000000100006488 nuttx`nx_start + 84
    frame #3: 0x00000001000042f4 nuttx`sim_gpio_initialize - 18446744069414567179
(lldb) c
Process 24321 resuming
Process 24321 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 2.1
    frame #0: 0x0000000100024584 nuttx`bzero
nuttx`bzero:
->  0x100024584 <+0>:  sub    sp, sp, #0x20
    0x100024588 <+4>:  stp    x29, x30, [sp, #0x10]
    0x10002458c <+8>:  str    x0, [sp, #0x8]
    0x100024590 <+12>: str    x1, [sp]
Target 0: (nuttx) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 2.1
  * frame #0: 0x0000000100024584 nuttx`bzero
    frame #1: 0x00000001000245a0 nuttx`bzero + 28
    frame #2: 0x00000001000245a0 nuttx`bzero + 28
    frame #3: 0x0000000100006488 nuttx`nx_start + 84
    frame #4: 0x00000001000042f4 nuttx`sim_gpio_initialize - 18446744069414567179
(lldb) q
```

It looks like bzero has recursive behavior on macos.Any good suggestions?

**Comment by anjiahao1** [comment]
@xiaoxiang781216 @masayuki2009 [pull11234](https://github.com/apache/nuttx/pull/11234) fix macos crash

**Comment by anjiahao1** [comment]
caused by [pr8651](https://github.com/apache/nuttx/pull/8651)
When compiling macos clang, bezero(s,n) and memset(s,0,n) are equivalent, causing recursion.

File: libs/libc/string/lib_bzero.c [hunk 3]
```diff
@@ -36,5 +36,21 @@
 
 void bzero(FAR void *s, size_t n)
```
Comment (xiaoxiang781216) [inline libs/libc/string/lib_bzero.c:37]: but can we change bzero to macro

**Comment by crafcat7** [comment]
[PR##11242 - Revert "libc/lib_bzero:Add bzero prototype."](https://github.com/apache/nuttx/pull/11242)
I reverted the original bzero function implementation to a macro

**Comment by xiaoxiang781216** [comment]
@anjiahao1 please drop the second patch

**Comment by masayuki2009** [comment]
> @anjiahao1 please drop the second patch

Did you test the latest PR on M1 Mac?

**Comment by anjiahao1** [comment]
> Did you test this PR with the latest master on M1 Mac?

image
yes i test on macbook with arm

**Pull request merged**